dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +253,282 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
         }
     }
 
-    void PushNotificationManager::RegisterActivator(PushNotificationActivationInfo const& details)
+    bool PushNotificationManager::IsBackgroundTaskRegistered(winrt::hstring const& backgroundTaskFullName)
     {
-        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Windows::PushNotifications::Feature_PushNotifications::IsEnabled());
+        auto tasks{ BackgroundTaskRegistration::AllTasks() };
+        bool isTaskRegistered{ std::any_of(std::begin(tasks), std::end(tasks),
+            [&](auto&& task)
+            {
+                auto name{ task.Value().Name() };
 
-        try
-        {
-            THROW_HR_IF_NULL(E_INVALIDARG, details);
+                if (std::wstring_view(name).substr(0, backgroundTaskName.size()) != backgroundTaskName)
+                {
+                    return false;
+                }
 
-            auto registrationActivators{ details.Activators() };
+                if (name == backgroundTaskFullName)
+                {
+                    m_pushTriggerRegistration = task.Value();
+                    return true;
+                }
 
-            auto isBackgroundTaskFlagSet{ WI_IsAnyFlagSet(registrationActivators, PushNotificationRegistrationActivators::PushTrigger | PushNotificationRegistrationActivators::ComActivator) };
-            THROW_HR_IF(E_INVALIDARG, isBackgroundTaskFlagSet && !IsActivatorSupported(registrationActivators));
+                throw winrt::hresult_invalid_argument(L"RegisterActivator has different clsid registered.");
+            }) };
 
-            auto isProtocolActivatorSet{ WI_IsFlagSet(registrationActivators, PushNotificationRegistrationActivators::ProtocolActivator) };
-            THROW_HR_IF(E_INVALIDARG, isProtocolActivatorSet && !IsActivatorSupported(registrationActivators));
+        return isTaskRegistered;
+    }
 
-            if (isProtocolActivatorSet)
+    void PushNotificationManager::Register()
+    {
+        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Windows::PushNotifications::Feature_PushNotifications::IsEnabled());
+
+        try
+        {
+            if (!AppModel::Identity::IsPackagedProcess())
             {
                 {
-                    auto lock = s_activatorInfoLock.lock_shared();
-                    THROW_HR_IF(E_INVALIDARG, s_protocolRegistration);
+                    auto lock = m_lock.lock_shared();
+                    THROW_HR_IF(E_INVALIDARG, m_protocolRegistration);
```

Comment:
You need to move THROW_HR_IF_MSG(E_FAIL, m_protocolRegistration, "The app has already been registered once!"); to the exclusive lock below. Otherwise you have introduced a race on this bool.